### ISSUE TITLE
ResNext

### ISSUE THREAD

**Issue by bhack** [endernewton/tf-faster-rcnn#27]
Do you plan to experiment with [ResNext](https://github.com/wenxinxu/ResNeXt-in-tensorflow)?

**Comment by endernewton** [comment]
I am still debugging ResNet. I will think about the direction after getting
done with that and see how much the AP can reach on COCO. However, ResNext
is probably not the next step I want to explore because it seems not help
AP much (in the original paper), I would probably start with FPN first if I
want to have better AP.

---------------------
Ender X. Chen

Language Technologies Institute,
School of Computer Science,
Carnegie Mellon University, Pittsburgh, PA

**Comment by bhack** [comment]
You can see ablative of ResNext in https://arxiv.org/format/1703.06870

**Comment by bhack** [comment]
See Table n. 3

**Comment by endernewton** [comment]
Oh yes I saw that coming out, it uses additional supervision from masks as
well. Not an exact apple-to-apple comparison for bbox annotation. The old
number in their original paper is very close for Resnet 101. The variation
of AP on COCO is smaller but on VOC the variation with different random
seeds can be as high as 1%.

---------------------
Ender X. Chen

Language Technologies Institute,
School of Computer Science,
Carnegie Mellon University, Pittsburgh, PA

**Comment by bhack** [comment]
Seems that they have analyzed distinc contributions: using RoIAlign (+1.1 APbb), multitask training (+0.9 APbb), and ResNeXt-101 (+1.6 APbb). All top performing backbones are using FPN.

**Comment by endernewton** [comment]
Yes, but the 1.6 addition is on top of the multi-task learning, 38.2->39.8,
what I mean is, they do not give an improvement of ResNext alone: how much
does it gain from 36.2/37.3 the FPN numbers (with or without RoIAlign).

Also I think RoIAlign is just crop and resize from tensorflow? Any idea?

---------------------
Ender X. Chen

Language Technologies Institute,
School of Computer Science,
Carnegie Mellon University, Pittsburgh, PA

**Comment by bhack** [comment]
The problem is that we don't have an explicit entry in Table 3 for Faster R-CNN, RoIAlign with ResNext FPN backbone to compare. But from the table just Faster R-CNN with ResNet and RoiAlign (without Multitask) seems to improve. RoiAlign is defined in "detail" in section 3. and seems that use bilinear interpolation.

**Comment by endernewton** [comment]
They say they will release the code over summer. I believe it will be in
caffe2. So we will see. :)

Anyway crop and resize does do better in my experiment, and it uses
bilinear interpolation.

---------------------
Ender X. Chen

Language Technologies Institute,
School of Computer Science,
Carnegie Mellon University, Pittsburgh, PA

**Comment by bhack** [comment]
Fair group.. probably Torch/Pytorch

**Comment by endernewton** [comment]
Yeah anything but tensorflow.

**Comment by bhack** [comment]
Figure 5 table (C) compare align and interpolation contribution of different pooling layers.

**Comment by philokey** [comment]
@endernewton I think RoIAlign may be has some different with crop_and_resize. RoiAlign not only use bilinear interpolation, it also use x/16 instead of [x/16].  However, you use tf.ceil to calculate the new width and height for crop_and_resize, I think it will lead to quantizes.

**Comment by endernewton** [comment]
@philokey actually I am using x/16.. the ceil thing is to measure the size of the entire conv5 feature map because crop_and_resize only takes relative size.. hmm now you actually reminds me that maybe the feature map in resnet101 can be different from vgg16. Needs some verification there, or maybe i should just take the size of the featuremap directly so it looks more clear.

**Comment by philokey** [comment]
@endernewton Why do you use ` ceil(x / 16 - 1) * 16`, when `x % 16 == 0`, it is seems wrong.

**Comment by endernewton** [comment]
@philokey "x" there is height or width factor for the entire image, not the locations.. please double check.

**Comment by bhack** [comment]
Just to notify https://github.com/CharlesShang/FastMaskRCNN /cc @CharlesShang

**Comment by endernewton** [comment]
Thanks for the pointer, if you are interested, you can try to incorporate that and submit a pull request. Implementing Mask RCNN is not in my bucket list, at least for now.

**Comment by bhack** [comment]
Could you be interested for the pyramid https://github.com/CharlesShang/FastMaskRCNN/blob/master/libs/nets/pyramid_network.py?

**Comment by endernewton** [comment]
I will do that in some time. As mentioned above, once I am done with ResNet part, I will probably implement FPN if further improving AP is needed. Implementing that will likely take more time because their official code is not released, so need to bug the authors if we want to clarify the details. So it will take some more rounds.

**Comment by zacwellmer** [comment]
I have _attempted_ to implement fpn with res50. I'm currently struggling to use the res50 imagenet weights to initialize the network. If anyone gets a chance the repo is [here](https://github.com/zacwellmer/tf-faster-rcnn/tree/pyramid). Let me know if you have any suggestions.

**Comment by lichengunc** [comment]
why do we need "-1" for the relative location computation? 
i.e., height = (conv5_height - 1)*16, then x = x/16

**Comment by chang010453** [comment]
> I have _attempted_ to implement FPN with res50. I'm currently struggling to use the res50 imagenet weights to initialize the network. If anyone gets a chance the repo is [here](https://github.com/zacwellmer/tf-faster-rcnn/tree/fpn), but nearly all the changes took place in [pyramid.py](https://github.com/zacwellmer/tf-faster-rcnn/blob/fpn/lib/nets/pyramid.py) Let me know if you have any suggestions.

Hi, have you finished your implement of FPN?
